### DISCUSSION TITLE
Example of integration with other "external" state managers (Zustand, Jotai, Valtio, Redux, etc)

### DISCUSSION THREAD

**Discussion by isBatak** [47ng/nuqs#683] [General]
Is there already any examples of integrations with Zustand or similar external state managers?

> integration with other "external" state managers (Zustand, Jotai, Valtio, Redux, etc)
https://github.com/47ng/nuqs/issues/375

cc @franky47

**Comment by franky47** [comment]
Thanks for opening this discussion!

There aren't any examples yet, I started experimenting with Jotai (which has an atom that reacts to URL changes, but isn't type-safe), and overall it feels like the path to integration could be:
1. A method for reading & parsing the search params as a one-off thing
2. A way to know when search params have been updated, to recompute any store items that may depend upon them.

Jotai has this reactivity concept with Atoms, but in traditional stores like Zustand & Redux, I'm not sure how that would look for computed states that depend on external reactivity sources.

Some concrete examples of external state management usage that depends on search params would be great to have a better understanding on how to compose them with nuqs.

**Reply by isBatak** [reply]
Both options, 1 and 2, would be a great addition to the library. They will allow users to manually connect external libraries like Zustand to the query state. 

Zustand has an example that describes how to persist external state in query search parameters: [Zustand URL Parameters Example](https://zustand.docs.pmnd.rs/guides/connect-to-state-with-url-hash#persist-and-connect-state-with-url-parameters-(example:-url-query-parameters)).

Regarding Zustand's reactivity, you can always manually update the state like this: `useFishStore.getState().increaseFishes()`.

Here is my example of external state management use-case:

1. The user clicks on an item, the app saves it to the Zustand store, which persists it to LocalStorage, and we set it as the active item. Currently, we are not saving this information in the URL, but we want to do it using nuqs.
2. If the user returns to the same page, we read from LocalStorage and set the default Zustand value.
3. The third case is when we want to redirect the user from another page to a specific "active" item. We want to transfer this information via a URL query parameter, and this is where nuqs comes into play.

It's also acceptable if in the 1. step we update the URL search param.

**Comment by leedia-tech** [comment]
A Zustand example would be very useful

**Reply by franky47** [reply]
Do you have a use-case in mind?

**Reply by flaviolivolsi** [reply]
Zustand suggests [this method](https://zustand.docs.pmnd.rs/guides/connect-to-state-with-url-hash) to sync the state with the URL, which works well but the outcome is quite ugly unfortunately. It would be nice to use the nuqs way with Zustand

**Comment by Jaimish00** [comment]
Hello, Any update on this? Can we get an example of using this with Zustand or Jotai? 
I am looking forward to this addition

**Comment by dgknca** [comment]
Hey, I have an advanced use-case that combines `atom`, `atomWithStorage`, `atomFamily` and nuqs. I couldn't find a proper way to use with `useQueryState`

**Comment by franky47** [comment]
I've been tinkering with Zustand, and loaders are a good way to inject URL state into the body of an "action" (function defined in the state):

https://github.com/user-attachments/assets/5a143202-0356-4000-9f25-c914fb333b6f

```tsx
'use client'

import { createLoader, parseAsFloat, useQueryStates } from 'nuqs'
import { Suspense } from 'react'
import { create } from 'zustand'

export default function Page() {
  return (
    <Suspense>
      <Client />
    </Suspense>
  )
}

const searchParams = {
  factor: parseAsFloat.withDefault(1)
}
const loadSearchParams = createLoader(searchParams)

function Client() {
  const [{ factor }, setSearchParams] = useQueryStates(searchParams)
  const bears = useBear(state => state.bears)
  const increasePopulation = useBear(state => state.increasePopulation)
  const removeAllBears = useBear(state => state.removeAllBears)
  return (
    <>
      <label style={{ display: 'flex', alignItems: 'center', gap: 8 }}>
        Factor:
        <input
          type="range"
          min={0.5}
          max={2}
          step={0.1}
          value={factor}
          onChange={e => setSearchParams({ factor: e.target.valueAsNumber })}
        />{' '}
        {factor}
      </label>
      <hr />
      <div>
        <button onClick={increasePopulation}>Increase Bears</button>
        <button onClick={removeAllBears}>Remove All Bears</button>
      </div>
      <p>Bears: {bears}</p>
    </>
  )
}

type BearState = {
  bears: number
  increasePopulation: () => void
  removeAllBears: () => void
}

const useBear = create<BearState>(set => ({
  bears: 0,
  increasePopulation: () => {
    const { factor } = loadSearchParams(location.search)
    set(state => ({ bears: state.bears + factor }))
  },
  removeAllBears: () => set({ bears: 0 })
}))
```

The main drawback I see with this is that it reads from the URL directly, and won't use the optimistic state that nuqs sets when writing to the URL (eg: when debouncing URL updates). But since that state is locked in React land, I don't see how to get to it without a major refactor to a single store, which is something that is planned and would likely make syncing against external stores easier.

To be fair, if you want to pass optimistic state into a Zustand action, it could be passed as an argument.
